File: vault-ios/VaultCommon/Features/RepoFilesDetails/RepoFilesDetailsScreenViewModel.swift
import Foundation
import VaultMobile

public class RepoFilesDetailsScreenViewModel: ObservableObject, WithRepoGuardViewModel {
    public let container: Container
    public let navController: MainNavController
    public let repoId: String
    public let encryptedPath: String

    public var detailsId: UInt32

    @Published public var content: RepoFilesDetailsScreenContent

    @Published public var info: Subscription<RepoFilesDetailsInfo>
    @Published public var file: Subscription<RepoFile>

    @Published public var repoGuardViewModel: RepoGuardViewModel

    @Published public var textEditorText: String

    private var shouldDestroyHandled: Bool

    public init(
        container: Container, navController: MainNavController, repoId: String,
        encryptedPath: String, isEditing: Bool
    ) {
        self.container = container
        self.navController = navController
        self.repoId = repoId
        self.encryptedPath = encryptedPath

        let detailsId = container.mobileVault.repoFilesDetailsCreate(
            repoId: repoId, encryptedPath: encryptedPath, isEditing: isEditing,
            options: RepoFilesDetailsOptions(
                loadContent: FilesFilter(categories: [.code, .text], exts: []),
                autosaveIntervalMs: 20000))

        self.detailsId = detailsId

        self.content = .loading

        info = Subscription(
            mobileVault: container.mobileVault,
            subscribe: { v, cb in
                v.repoFilesDetailsInfoSubscribe(detailsId: detailsId, cb: cb)
            },
            getData: { v, id in
                v.repoFilesDetailsInfoData(id: id)
            })

        file = Subscription(
            mobileVault: container.mobileVault,
            subscribe: { v, cb in
                v.repoFilesDetailsFileSubscribe(detailsId: detailsId, cb: cb)
            },
            getData: { v, id in
                v.repoFilesDetailsFileData(id: id)
            })

        repoGuardViewModel = RepoGuardViewModel(
            container: container, repoId: repoId, setupBiometricUnlockVisible: true)

        textEditorText = ""

        shouldDestroyHandled = false

        self.info.setOnData { [weak self] data in
            if let self = self {
                if let info = data {
                    self.repoGuardViewModel.update(
                        repoStatus: info.repoStatus, isLocked: info.isLocked)

                    if info.shouldDestroy && !shouldDestroyHandled {
                        shouldDestroyHandled = true

                        if let repoId = info.repoId {
                            let routes = info.encryptedParentPathChain.map {
                                MainRoute.repoFiles(repoId: repoId, encryptedPath: $0)
                            }

                            DispatchQueue.main.async {
                                self.navController.replace(routes)
                            }
                        }
                    }
                }
            }
        }

        self.file.setOnData { [weak self] data in
            if let self = self {
                if let file = data {
                    self.load(file)
                }
            }
        }
    }

    deinit {
        container.mobileVault.repoFilesDetailsDestroy(detailsId: detailsId)
    }

    public func retryLoad() {
        if let file = file.data {
            load(file)
        }
    }

    private func load(_ file: RepoFile) {
        if RepoFilesDetailsScreenContentData.isTextEditor(file) {
            content = .textEditor(file: file)

            return
        }

        if let loader = RepoFilesDetailsScreenContentData.getLoader(
            file: file,
            onWarning: { warning in
                self.container.mobileVault.notificationsShow(message: warning)
            })
        {
            content = .downloading

            let localBasePath = container.storageHelper.getTempDir().path

            container.mobileVault.repoFilesDetailsDownloadTempFile(
                detailsId: detailsId,
                localBasePath: localBasePath,
                onDone: TransfersDownloadDoneFn { [weak self] localFilePath in
                    if let self = self {
                        let localFileURL = URL(fileURLWithPath: localFilePath)

                        Task { @MainActor in
                            let data = await loader(localFileURL)

                            self.content = .downloaded(localFileURL: localFileURL, data: data)
                        }
                    }
                }
            )
        } else {
            content = .notSupported(file: file)
        }
    }
}

File: vault-android/app/src/main/java/net/koofr/vault/features/repofilesdetails/RepoFilesDetailsScreenViewModel.kt
package net.koofr.vault.features.repofilesdetails

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import androidx.compose.runtime.mutableStateOf
import androidx.core.content.FileProvider
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.navigation.NavController
import coil.ImageLoader
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.launch
import net.koofr.vault.BuildConfig
import net.koofr.vault.FileCategory
import net.koofr.vault.FilesFilter
import net.koofr.vault.MobileVault
import net.koofr.vault.RepoFile
import net.koofr.vault.RepoFilesDetailsOptions
import net.koofr.vault.TransfersDownloadDone
import net.koofr.vault.features.downloads.DownloadHelper
import net.koofr.vault.features.fileicon.FileIconCache
import net.koofr.vault.features.mobilevault.Subscription
import net.koofr.vault.features.repo.RepoGuardViewModel
import net.koofr.vault.features.repo.WithRepoGuardViewModel
import net.koofr.vault.features.storage.StorageHelper
import java.io.File
import javax.inject.Inject

@HiltViewModel
class RepoFilesDetailsScreenViewModel @Inject constructor(
    val mobileVault: MobileVault,
    val fileIconCache: FileIconCache,
    private val storageHelper: StorageHelper,
    private val downloadHelper: DownloadHelper,
    private val imageLoader: ImageLoader,
    savedStateHandle: SavedStateHandle,
    @SuppressLint("StaticFieldLeak") @ApplicationContext private val appContext: Context,
) : ViewModel(), WithRepoGuardViewModel {
    private var repoGuardViewModel: RepoGuardViewModel? = null

    private val repoId: String = savedStateHandle.get<String>("repoId")!!
    private val encryptedPath: String = savedStateHandle.get<String>("path")!!

    val detailsId = mobileVault.repoFilesDetailsCreate(
        repoId = repoId,
        encryptedPath = encryptedPath,
        isEditing = false,
        options = RepoFilesDetailsOptions(
            loadContent = FilesFilter(
                categories = listOf(FileCategory.CODE, FileCategory.TEXT),
                exts = emptyList(),
            ),
            autosaveIntervalMs = 20000u,
        ),
    ).also {
        addCloseable {
            mobileVault.repoFilesDetailsDestroy(detailsId = it)
        }
    }

    val info = Subscription(
        mobileVault = mobileVault,
        coroutineScope = viewModelScope,
        subscribe = { v, cb -> v.repoFilesDetailsInfoSubscribe(detailsId = detailsId, cb = cb) },
        getData = { v, id ->
            v.repoFilesDetailsInfoData(id = id).also {
                it?.let {
                    repoGuardViewModel?.update(it.repoStatus, it.isLocked)
                }
            }
        },
    ).also {
        addCloseable(it)
    }

    var currentFile: RepoFile? = null

    val content =
        mutableStateOf<RepoFilesDetailsScreenContent>(RepoFilesDetailsScreenContent.Loading)

    override fun onCleared() {
        super.onCleared()

        content.value.close()
    }

    override fun setRepoGuardViewModel(repoGuardViewModel: RepoGuardViewModel) {
        if (this.repoGuardViewModel != null) {
            return
        }

        this.repoGuardViewModel = repoGuardViewModel

        addCloseable {
            this.repoGuardViewModel = null
        }

        info.data.value?.let {
            repoGuardViewModel.update(it.repoStatus, it.isLocked)
        }
    }

    fun setContent(newContent: RepoFilesDetailsScreenContent) {
        content.value.close()

        content.value = newContent
    }

    fun load(file: RepoFile) {
        currentFile = file

        val loader = RepoFilesDetailsScreenContentData.getLoader(appContext, file, imageLoader)

        if (loader != null) {
            setContent(RepoFilesDetailsScreenContent.Downloading)

            mobileVault.repoFilesDetailsDownloadTempFile(
                detailsId = detailsId,
                localBasePath = storageHelper.getTempDir(),
                onDone = object : TransfersDownloadDone {
                    override fun onDone(localFilePath: String, contentType: String?) {
                        viewModelScope.launch {
                            val localFile = File(localFilePath)

                            // prevent loading race conditions
                            if (currentFile == file) {
                                val data = loader(localFile)

                                setContent(
                                    RepoFilesDetailsScreenContent.Downloaded(
                                        file,
                                        localFile,
                                        data,
                                    ),
                                )
                            }
                        }
                    }
                },
            )
        } else {
            setContent(RepoFilesDetailsScreenContent.NotSupported(file))
        }
    }

    fun share(context: Context, localFile: File, contentType: String?) {
        val uri = FileProvider.getUriForFile(context, BuildConfig.FILES_AUTHORITY, localFile)

        val intent = Intent().apply {
            action = Intent.ACTION_SEND
            putExtra(Intent.EXTRA_STREAM, uri)
            contentType?.let {
                type = it
            }
        }

        context.startActivity(Intent.createChooser(intent, null))
    }

    fun download(navController: NavController, file: RepoFile) {
        downloadHelper.downloadRepoFile(navController, file)
    }

    fun rename() {
        info.data.value?.let { info ->
            info.repoId?.let { repoId ->
                info.encryptedPath?.let { encryptedPath ->
                    mobileVault.repoFilesRenameFile(repoId = repoId, encryptedPath = encryptedPath)
                }
            }
        }
    }

    fun delete() {
        mobileVault.repoFilesDetailsDelete(detailsId = detailsId)
    }
}
